docs/libvirt-node-networking.md

```diff
@@ -0,0 +1,298 @@
+DEV_SUBNET=$(echo "$DEV_IP" | cut -d"." -f1-3)
+
+# Create a host ipvlan through which the host and the guests can talk to each other
+ip link add macvlan0 link $DEV type macvlan mode bridge
+ip link set dev macvlan0 up
+
+# here we multiple options:
+# * use dhcp, it will do the default routes for us
+# * do dhcp without setting the gateway, we then have to set it manually
+# * unconfigure $DEV and set the IP on macvlan0 and manully change the default route
+# using an extra IP via dhcp sounds like the best solution.
+# We can leave all other routes which belong to $DEV untouched.
+# No matter what we do, we don't want any routes modified.
+ip addr add $DEV_SUBNET.13/24 dev macvlan0 noprefixroute
+
+# normally we already have the netork namespace of libvirt given.
+ip netns add ns1
+
+# create the macvlan link attaching it to the parent host $DEV
+ip link add macvlan1 link $DEV type macvlan mode bridge
+
+# move the new interface macvlan1 to the new namespace
+ip link set macvlan1 netns ns1
+
+# bring the interface up
+# ns1 will later be used by libvirt to attach macvtaps to
+ip netns exec ns1 ip link set dev macvlan1 up
+```
+
+After that invocation, we have a host macvlan, with an IP address assigned. The
+CNI dhcp plugin comes with a minimalistic dhcp client, which we can use to do
+the DHCP renewals. An important side note here is, that the minimalistic daemon
+returns the acquired IP alog side with routing rules suggested by the DHCP
+server, but they are not applied, so the host is not modified.
+
+#### Prepare the VM device
+
+When a VM contains a network configuration like this:
+
+```yaml
+kind: VM
+spec:
+  domain:
+    devices:
+      interfaces:
+      - type: HostNetwork
+        mac:
+          address: 00:11:22:33:44:55
+        model:
+          type: virtio
+```
+For every `HostNetwork`, `virt-handler` let's libvirt create a new macvtap
+attached to the one prepared macvlan in the initial setup step in the libvirt
+namespace.  Ee will use the CNI `dhcp` IPAM plugin with a provided or generated
+MAC address. Based on this MAC address and the IPAM plugin invocation we end up
+with the correct IP and MAC pair. The MAC address will then be used by libvirt
+as the mactap address. The IP can be used to add a route, to allow routing
+between the new guest and the host.
+
+Taking the macvlan name and the MAC address, we can now transform the
+HostNetwork entry into a libvirt XML fragment:
+
+```xml
+<interface type='direct'>
+  <mac address="fe:54:00:f7:78:6d"/>
+  <source dev='macvlan1' mode='bridge'/>
+</interface>
+```
+
+When the guest now tries to acquire an IP address with this mac address, it
+will get the same IP address. Based on that assumption we can create the route
+to allow the guest to talk to the node.
+
+```bash
+# Add an extra route which allows accessing the $DEV_IP through the master macvlan.
+# We need one of these routes per device.
+ip route add $DEV_SUBNET.11/32 via $DEV_SUBNET.13 dev macvlan0
```

Comment:
I'm missing something here: This is a route to the host IP via the master macvlan.
But what will cause the opposite direction to pass through the master macvlan?
Issuing a pint from the host to the VM IP... There is no route that requires the packets to exit the master macvlan, it may exit the host interface instead.

Reply:
Adding that  route makes it work for both directions for me (although I had a the default routes going through my master device). I will try to find a definite answer, why I did not need it, or add the additional route, if the described scenario can occur.